org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/FinallyFilter.java

```diff
@@ -0,0 +1,196 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2017 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis.filter;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.JumpInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.TryCatchBlockNode;
+import org.objectweb.asm.tree.VarInsnNode;
+
+/**
+ * Filters duplicates of finally blocks that compiler generates.
+ */
+public final class FinallyFilter implements IFilter {
+
+	public void filter(final String className, final String superClassName,
+			final MethodNode methodNode, final IFilterOutput output) {
+		for (final TryCatchBlockNode tryCatchBlock : methodNode.tryCatchBlocks) {
+			if (tryCatchBlock.type == null) {
+				new Matcher(output, methodNode, tryCatchBlock).filter();
+			}
+		}
+	}
+
+	private static class Matcher {
```

Comment:
@Godin The only state in `Matcher` class is member `inside`, which is used only locally within method `filter()`. Therefore should be better al local variable within this method. Then the `Matcher` class can be removed at all. Tried this here: https://gist.github.com/marchof/5adb1930188136c3ea340ff430a2cac5

Reply:
@marchof done

Reply:
@Godin Thanks! :white_check_mark: